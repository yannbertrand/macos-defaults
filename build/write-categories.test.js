const fs = require('fs')
jest.mock('fs')

const writeCategories = require('./write-categories')
require('./handlebars-helpers')

const templatesPath = 'templates'
const destinationPath = 'dist'

describe('write-categories', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('one category, no page', () => {
    describe('with image', () => {
      beforeEach(() =>
        callWriteCategories({
          categories: [
            {
              folder: 'category',
              name: 'Category',
              description: 'This is a description of the category. It can contain "quotes" and markdown.\n\n- a list item',
              image: {
                filename: 'category.png',
                width: 740,
                height: 80,
              },
            },
          ],
        })
      )

      it('should write a category/readme.md file using the category template', () => {
        const categoryReadmeContent = readFile(
          `${destinationPath}/category/readme.md`
        )
        expect(categoryReadmeContent).toMatchSnapshot()
      })

      it('should copy the category.png example image to a category folder', () => {
        const originImagePath = readFile(
          `${destinationPath}/category/category.png`
        )
        expect(originImagePath).toMatchInlineSnapshot(
          `"copied:../images/category/category.png"`
        )
      })
    })
  })

  describe('one category, one page', () => {
    beforeEach(() =>
      callWriteCategories({
        categories: [
          {
            folder: 'category',
            name: 'Category',
            description: 'Category description.',
            keys: [
              {
                key: 'page',
                domain: 'com.apple.category',
                title: 'Page',
                description: 'Page description.',
                param: { type: 'bool' },
                examples: [
                  {
                    value: '~/Desktop',
                    default: true,
                    text: 'output when value is ~/Desktop',
                  },
                  {
                    value: '~/Pictures',
                    text: 'output when value is ~/Pictures',
                  },
                ],
                versions: ['Big Sur'],
              },
            ],
          },
        ],
      })
    )

    it('should write a category/readme.md file using the category template', () => {
      const categoryReadmeContent = readFile(
        `${destinationPath}/category/readme.md`
      )
      expect(categoryReadmeContent).toMatchSnapshot()
    })
  })

  describe('one category, two pages', () => {
    beforeEach(() =>
      callWriteCategories({
        categories: [
          {
            folder: 'category',
            name: 'Category',
            description: 'Category description.',
            keys: [
              {
                key: 'page1',
                domain: 'com.apple.category',
                title: 'Page 1',
                description: 'Page 1 description.',
                param: { type: 'bool' },
                examples: [
                  {
                    value: true,
                    default: true,
                    text: 'output when value is true',
                  },
                  {
                    value: false,
                    text: 'output when value is false',
                  },
                ],
                versions: ['Big Sur'],
              },
              {
                key: 'page2',
                domain: 'com.apple.category',
                title: 'Page 2',
                description: 'Page 2 description.',
                param: { type: 'bool' },
                examples: [
                  {
                    value: true,
                    text: 'output when value is true',
                  },
                  {
                    value: false,
                    default: true,
                    text: 'output when value is false',
                  },
                ],
                versions: ['Big Sur'],
              },
            ],
          },
        ],
      })
    )

    it('should write a category/readme.md file using the category template', () => {
      const categoryReadmeContent = readFile(
        `${destinationPath}/category/readme.md`
      )
      expect(categoryReadmeContent).toMatchSnapshot()
    })
  })

  describe('two categories, one page in each', () => {
    beforeEach(() =>
      callWriteCategories({
        categories: [
          {
            folder: 'category1',
            name: 'Category 1',
            description: 'Category 1 description.',
            keys: [
              {
                key: 'page',
                domain: 'com.apple.category1',
                title: 'Page',
                description: 'Page description.',
                param: { type: 'bool' },
                examples: [
                  {
                    value: true,
                    default: true,
                    text: 'output when value is true',
                  },
                  {
                    value: false,
                    text: 'output when value is false',
                  },
                ],
                versions: ['Big Sur'],
              },
            ],
          },
          {
            folder: 'category2',
            name: 'Category 2',
            description: 'Category 2 description.',
            keys: [
              {
                key: 'page',
                domain: 'com.apple.category2',
                title: 'Page',
                description: 'Page description.',
                param: { type: 'bool' },
                examples: [
                  {
                    value: true,
                    text: 'output when value is true',
                  },
                  {
                    value: false,
                    default: true,
                    text: 'output when value is false',
                  },
                ],
                versions: ['Big Sur'],
              },
            ],
          },
        ],
      })
    )

    it('should write a category1/readme.md file using the category template', () => {
      const categoryReadmeContent = readFile(
        `${destinationPath}/category1/readme.md`
      )
      expect(categoryReadmeContent).toMatchSnapshot()
    })

    it('should write a category2/readme.md file using the category template', () => {
      const categoryReadmeContent = readFile(
        `${destinationPath}/category2/readme.md`
      )
      expect(categoryReadmeContent).toMatchSnapshot()
    })
  })
})

const callWriteCategories = (defaults) =>
  writeCategories(defaults, templatesPath, destinationPath)
const readFile = (file) => fs.readFakeFileSync(file, 'utf8')
