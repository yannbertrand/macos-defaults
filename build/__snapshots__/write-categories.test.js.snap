// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`write-categories one category, no page with image should write a category/readme.md file using the category template 1`] = `
"---
metaTitle: Category | macOS defaults
meta:
  - property: "og:title"
    content: macOS defaults > Category
  - name: "description"
    content: "This is a description of the category. It can contain &quot;quotes&quot; and markdown.\\n\\n- a list item"
  - property: "og:description"
    content: "This is a description of the category. It can contain &quot;quotes&quot; and markdown.\\n\\n- a list item"
  - property: "twitter:card"
    content: "summary"
  - property: "twitter:image"
    content: "https://macos-defaults.netlify.app/media-1x1.webp"
  - property: "og:image"
    content: "https://macos-defaults.netlify.app/media-1x1.jpg"
---
# Category

<img
  src="./category.png" alt="A screenshot of the Category"
  width="740" height="80" style="height: auto"
/>

This is a description of the category. It can contain "quotes" and markdown.

- a list item

## Keys

"
`;

exports[`write-categories one category, one page should write a category/readme.md file using the category template 1`] = `
"---
metaTitle: Category | macOS defaults
meta:
  - property: "og:title"
    content: macOS defaults > Category
  - name: "description"
    content: "Category description."
  - property: "og:description"
    content: "Category description."
  - property: "twitter:card"
    content: "summary"
  - property: "twitter:image"
    content: "https://macos-defaults.netlify.app/media-1x1.webp"
  - property: "og:image"
    content: "https://macos-defaults.netlify.app/media-1x1.jpg"
---
# Category

Category description.

## Keys

- [Page](./page.html)
"
`;

exports[`write-categories one category, two pages should write a category/readme.md file using the category template 1`] = `
"---
metaTitle: Category | macOS defaults
meta:
  - property: "og:title"
    content: macOS defaults > Category
  - name: "description"
    content: "Category description."
  - property: "og:description"
    content: "Category description."
  - property: "twitter:card"
    content: "summary"
  - property: "twitter:image"
    content: "https://macos-defaults.netlify.app/media-1x1.webp"
  - property: "og:image"
    content: "https://macos-defaults.netlify.app/media-1x1.jpg"
---
# Category

Category description.

## Keys

- [Page 1](./page1.html)
- [Page 2](./page2.html)
"
`;

exports[`write-categories two categories, one page in each should write a category1/readme.md file using the category template 1`] = `
"---
metaTitle: Category 1 | macOS defaults
meta:
  - property: "og:title"
    content: macOS defaults > Category 1
  - name: "description"
    content: "Category 1 description."
  - property: "og:description"
    content: "Category 1 description."
  - property: "twitter:card"
    content: "summary"
  - property: "twitter:image"
    content: "https://macos-defaults.netlify.app/media-1x1.webp"
  - property: "og:image"
    content: "https://macos-defaults.netlify.app/media-1x1.jpg"
---
# Category 1

Category 1 description.

## Keys

- [Page](./page.html)
"
`;

exports[`write-categories two categories, one page in each should write a category2/readme.md file using the category template 1`] = `
"---
metaTitle: Category 2 | macOS defaults
meta:
  - property: "og:title"
    content: macOS defaults > Category 2
  - name: "description"
    content: "Category 2 description."
  - property: "og:description"
    content: "Category 2 description."
  - property: "twitter:card"
    content: "summary"
  - property: "twitter:image"
    content: "https://macos-defaults.netlify.app/media-1x1.webp"
  - property: "og:image"
    content: "https://macos-defaults.netlify.app/media-1x1.jpg"
---
# Category 2

Category 2 description.

## Keys

- [Page](./page.html)
"
`;
